File: src/blacklist/blacklist.php
<?php

/**
 * Load blacklist IPs from a blacklist configuration file.
 *
 * Uses `read_file()` and `extractIPs()` helpers. Returns an array of IPs
 * (empty array if the file is missing or contains no IPs).
 *
 * @param string|null $blacklistConf Optional path to a blacklist configuration file
 * @return array Array of IP strings
 */
function get_blacklist($blacklistConf = null)
{
  $path = !empty($blacklistConf) ? $blacklistConf : __DIR__ . '/../../data/blacklist.conf';
  if (!is_string($path) || !is_file($path) || !is_readable($path)) {
    return [];
  }

  $maxCidrExpansion = 65536;
  $results          = [];
  $handle           = fopen($path, 'rb');
  if ($handle === false) {
    return [];
  }

  while (($line = fgets($handle)) !== false) {
    $line = trim($line);
    if (empty($line) || strpos($line, '#') === 0) {
      continue;
    }

    // check if line is CIDR notation
    if (strpos($line, '/') !== false) {
      // Only expand IPv4 CIDR here; IPv6 CIDR expansion can be too large.
      if (preg_match('/^(?:\d{1,3}\.){3}\d{1,3}\/(?:[0-9]|[12][0-9]|3[0-2])$/', $line) === 1) {
        $cidrParts = explode('/', $line, 2);
        if (!is_array($cidrParts) || count($cidrParts) !== 2) {
          continue;
        }

        $cidrIp   = trim($cidrParts[0]);
        $cidrMask = trim($cidrParts[1]);
        if (!filter_var($cidrIp, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)) {
          continue;
        }
        if (!is_numeric($cidrMask)) {
          continue;
        }

        $cidrMaskInt = (int) $cidrMask;
        if ($cidrMaskInt < 0 || $cidrMaskInt > 32) {
          continue;
        }

        $cidrExpansionSize = (int) pow(2, (32 - $cidrMaskInt));
        if ($cidrExpansionSize > $maxCidrExpansion) {
          continue;
        }

        $packedIp = inet_pton($cidrIp);
        if ($packedIp !== false) {
          $unpacked = unpack('N', $packedIp);
          if (is_array($unpacked) && isset($unpacked[1])) {
            $ipInt = (int) $unpacked[1];
            $mask  = $cidrMaskInt;

            $maskInt = $mask === 0 ? 0 : ((0xFFFFFFFF << (32 - $mask)) & 0xFFFFFFFF);
            $start   = $ipInt & $maskInt;
            $end     = $start | (~$maskInt & 0xFFFFFFFF);

            for ($current = $start; $current <= $end; $current++) {
              $ip = long2ip((int) $current);
              if (is_string($ip) && filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)) {
                $results[$ip] = true;
              }
              if ($current === 0xFFFFFFFF) {
                break;
              }
            }
          }
        }
      }
    } else {
      // check if line is a valid IP
      if (filter_var($line, FILTER_VALIDATE_IP)) {
        $results[$line] = true;
      }
    }
  }
  fclose($handle);

  return array_keys($results);
}

/**
 * Check if a proxy (IP or IP:PORT) is blacklisted.
 *
 * Accepts strings like `IP`, `IP:PORT`, `[IPv6]:PORT`, or raw IPv6.
 * Extracts the IP portion and checks it against the blacklist entries.
 *
 * @param string $proxy Proxy string (IP or IP:PORT or [IPv6]:PORT)
 * @param string|null $blacklistConf Optional path to a blacklist configuration file
 * @return bool True if the extracted IP is present in the blacklist
 */
function is_blacklist($proxy, $blacklistConf = null)
{
  if (empty($proxy) || !is_string($proxy)) {
    return false;
  }
  $p  = trim($proxy);
  $ip = null;

  // IPv6 in brackets: [::1]:8080
  if (preg_match('/^\[([^\]]+)\](?::\d+)?$/', $p, $m)) {
    $ip = $m[1];
  } else {
    // If whole string is a valid IP (v4 or v6)
    if (filter_var($p, FILTER_VALIDATE_IP)) {
      $ip = $p;
    } elseif (preg_match('/^(.+):\d+$/', $p, $m2)) {
      // Looks like IP:PORT (handles IPv6:PORT when IPv6 is not bracketed)
      $candidate = $m2[1];
      if (filter_var($candidate, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6) || filter_var($candidate, FILTER_VALIDATE_IP, FILTER_FLAG_IPV4)) {
        $ip = $candidate;
      }
    }
    // Fallback: extract first IPv4 occurrence
    if (empty($ip) && preg_match('/(\d{1,3}(?:\.\d{1,3}){3})/', $p, $m3)) {
      $ip = $m3[1];
    }
  }

  if (empty($ip)) {
    return false;
  }

  $blacklist = get_blacklist($blacklistConf);
  if (empty($blacklist) || !is_array($blacklist)) {
    return false;
  }
  return in_array($ip, $blacklist, true);
}

/**
 * Remove blacklisted IPs from the proxies database.
 *
 * This function delegates loading of blacklist entries to `get_blacklist()` and
 * deletes rows from `proxies` where `proxy` contains a blacklisted IP and
 * `status` is not `active`. Supports PDO drivers `sqlite` and `mysql`.
 *
 * @param PDO        $pdo           PDO instance connected to the proxies database
 * @param string|null $blacklistConf Optional path to a blacklist configuration file
 * @return void
 */
function blacklist_remover($pdo, $blacklistConf = null)
{
  $driver = $pdo->getAttribute(PDO::ATTR_DRIVER_NAME);
  if (!in_array($driver, ['sqlite', 'mysql'])) {
    echo "[BLACKLIST] Unsupported database driver: $driver. Skipping blacklist removal.\n";
    return;
  }

  $blacklist = get_blacklist($blacklistConf);
  if (empty($blacklist)) {
    return;
  }

  foreach ($blacklist as $ip) {
    $query = 'DELETE FROM proxies WHERE proxy LIKE :proxy AND status != :active_status';
    $stmt  = $pdo->prepare($query);

    $proxy        = "%$ip%";
    $activeStatus = 'active';
    $stmt->bindParam(':proxy', $proxy, PDO::PARAM_STR);
    $stmt->bindParam(':active_status', $activeStatus, PDO::PARAM_STR);

    if ($stmt->execute()) {
      $affectedRows = $stmt->rowCount();
      echo "[BLACKLIST] $ip deleted $affectedRows row(s).\n";
    } else {
      $err = $stmt->errorInfo();
      echo "[BLACKLIST] $ip failed to delete rows: " . json_encode($err) . "\n";
    }
  }
}

File: src/blacklist/blacklist.py
"""Blacklist helpers for loading and checking IP entries.

This module ports the behavior from `src/blacklist/blacklist.php`:
- `get_blacklist()` reads a blacklist file and returns unique IPs.
- IPv4 CIDR entries are expanded with a safety cap.
- `is_blacklist()` accepts IP, IP:PORT, [IPv6]:PORT, or raw IPv6.
"""

from __future__ import annotations

from ipaddress import IPv4Address, ip_address, ip_network
from proxy_hunter import extract_proxies
from pathlib import Path
import re

MAX_CIDR_EXPANSION = 65536
_IPV4_IN_TEXT_RE = re.compile(r"(\d{1,3}(?:\.\d{1,3}){3})")


def _default_blacklist_path() -> Path:
    """Return the default blacklist file path used by this project."""
    return Path(__file__).resolve().parents[2] / "data" / "blacklist.conf"


def get_blacklist(blacklist_conf: str | None = None) -> list[str]:
    """Load blacklist IPs from file.

    Args:
            blacklist_conf: Optional path to a blacklist config file.

    Returns:
            A list of unique IP strings. Returns an empty list if the file
            is missing, unreadable, or contains no valid entries.
    """
    path = Path(blacklist_conf) if blacklist_conf else _default_blacklist_path()
    if not path.is_file():
        return []

    results: dict[str, bool] = {}

    try:
        with path.open("r", encoding="utf-8", errors="ignore") as handle:
            for raw_line in handle:
                line = raw_line.strip()
                if not line or line.startswith("#"):
                    continue

                if "/" in line:
                    # Expand only IPv4 CIDR entries and skip very large ranges.
                    try:
                        network = ip_network(line, strict=False)
                    except ValueError:
                        continue

                    if network.version != 4:
                        continue

                    if network.num_addresses > MAX_CIDR_EXPANSION:
                        continue

                    for ip in network:
                        results[str(ip)] = True
                else:
                    try:
                        parsed = ip_address(line)
                    except ValueError:
                        continue
                    results[str(parsed)] = True
    except OSError:
        return []

    return list(results.keys())


def is_blacklist(proxy: str, blacklist_conf: str | None = None) -> bool:
    """Check whether an IP extracted from proxy text is blacklisted.

    Accepted proxy formats:
    - ``IP``
    - ``IP:PORT``
    - ``[IPv6]:PORT``
    - raw ``IPv6``
    """
    if not isinstance(proxy, str) or not proxy.strip():
        return False

    value = proxy.strip()
    ip: str | None = None

    # IPv6 in brackets: [::1]:8080 or [::1]
    bracket_match = re.match(r"^\[([^\]]+)\](?::\d+)?$", value)
    if bracket_match:
        ip = bracket_match.group(1)
    else:
        # Whole string as IP first.
        try:
            parsed = ip_address(value)
            ip = str(parsed)
        except ValueError:
            pass

        # Candidate from IP:PORT where the left side can be IPv4 or IPv6.
        if ip is None:
            port_match = re.match(r"^(.+):\d+$", value)
            if port_match:
                candidate = port_match.group(1)
                try:
                    parsed_candidate = ip_address(candidate)
                    ip = str(parsed_candidate)
                except ValueError:
                    pass

        # Fallback: first IPv4-like occurrence in text.
        if ip is None:
            fallback = _IPV4_IN_TEXT_RE.search(value)
            if fallback:
                ip = fallback.group(1)

    if ip is None:
        return False

    blacklist = get_blacklist(blacklist_conf)
    if not blacklist:
        return False

    return ip in blacklist


if __name__ == "__main__":
    proxies = "206.123.156.232:8080"
    ips = extract_proxies(proxies)
    for data in ips:
        ip = data.proxy.split(":")[0] if ":" in data.proxy else data.proxy
        print(f"Checking {ip} against blacklist...")
        if is_blacklist(ip):
            print(f"{ip} is blacklisted.")
        else:
            print(f"{ip} is not blacklisted.")
